src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -252,6 +252,38 @@ private class Emit(
       case ArrayLen(a) =>
         val (doa, ma, va) = emit(a)
         (doa, ma, TContainer.loadLength(region, coerce[Long](va)))
+      case x@ArrayRange(startir, stopir, stepir) =>
+        val (d, m, v) = Array(emit(startir), emit(stopir), emit(stepir)).unzip3
+        val start = fb.newLocal[Int]("ar_start")
+        val stop = fb.newLocal[Int]("ar_stop")
+        val step = fb.newLocal[Int]("ar_step")
+        val len = fb.newLocal[Int]("ar_len")
+        val srvb = new StagedRegionValueBuilder(fb, x.typ)
+        (coerce[Unit](Code(d: _*)), m.reduce(_ || _), Code(
+          start := coerce[Int](v(0)),
+          stop := coerce[Int](v(1)),
+          step := coerce[Int](v(2)),
+          step.ceq(0).mux(
+            Code._fatal("Array range cannot have step size 0."),
+            Code._empty[Unit]),
+          len := start.ceq(stop).mux(0,
+            (step < 0).mux((start - stop - 1) / (const(0) - step) + 1,
+              (stop - start - 1) / step + 1)),
+          len := (len < 0).mux(0, len),
+          srvb.start(len, init=true),
+          (step < 0).mux(
```

Comment:
I think it's cleaner to mux on the condition, see Array.range
```
    var i = start
    while (if (step < 0) end < i else i < end) {
      b += i
      i += step
    }
```

Reply:
I think so too, but then you need to check the `step < 0` condition every time you go through the loop. Happy to change it, either way.

Reply:
Good point, I think better to keep it as is then.